### PR TITLE
Allow async/slow implementations of authorization checks for Binder gRPCs.

### PR DESCRIPTION
**Pull request by mateusazis**
Introduce `AsyncSecurityPolicy`, exposing a method returns a `ListenableFuture<Status>` that callers can implement to perform slower auth checks (like network calls, disk I/O etc.) without necessarily blocking the gRPC calling thread.

Partially addresses: https://github.com/grpc/grpc-java/issues/10566

### CODE REVIEW

File: binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java
Comment (markb74) [inline binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java:48]: Thinking this through some more, I wonder if we should throw an UnsupportedOperationException error here instead. 

Otherwise we have no way of handling the possibility of a transient failure (which shouldn't be cached).

File: binder/src/main/java/io/grpc/binder/SecurityPolicy.java
Comment (markb74) [inline binder/src/main/java/io/grpc/binder/SecurityPolicy.java:46]: Let's leave this comment out for now, since AsyncSecurityPolicy will be experimental for a time.

We can enforce this preference internally since we'll able to handle any changes to the new API, but 3p users won't have the luxury.

**Review by markb74 — CHANGES_REQUESTED** [review]

File: binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java
Comment (markb74) [inline binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java:19]: @ejona86 is there anything special we should do with the build in case ListenableFuture is problematic part of the API for some users?

File: binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java]: Hard to say. This code won't really be used for long, only if someone starts using `AsyncSecurityPolicy` right as-is right after this PR is merged. On the next one, I'll make `TransportAuthorizationState.checkAuthorization` call the async method, so I believe there will no longer be any internal caller anymore (though, of course, integrators can always still call it, even though they shouldn't).

Also added a couple more unit tests to handle cancelled vs interrupted futures. I'm re-interrupting the current thread based on http://go/java-practices/interruption#interruptedexception.

Comment (markb74) [reply binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java]: Sorry to be annoying, but I think it'll be simpler if we just leave this file out of the CL for now, and only add it in the follow up when we also include support for it. Then the question of what we do about this method becomes much much easier.

Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java]: Done. I've also moved the new unit tests out since they required `AsyncSecurityPolicy`.

File: binder/src/main/java/io/grpc/binder/BinderInternal.java
Comment (markb74) [inline binder/src/main/java/io/grpc/binder/BinderInternal.java:37]: We can leave @ExperimentalApi out here since this class is internal.

File: binder/src/main/java/io/grpc/binder/BinderServerBuilder.java
Comment (markb74) [inline binder/src/main/java/io/grpc/binder/BinderServerBuilder.java:34]: Remove this line? (it's now a dupe below).

File: binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java
Comment (markb74) [inline binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java:66]: maybe "enforcer" -> "checker" since it's actually this class which enforces the policy.

**Review by markb74 — APPROVED** [review]
Thanks, just a couple of minor nits.

**Comment by markb74** [comment]
@jdcormie & @ejona86 - Since this is making a change to how we enforce security, can I ask you to give this a once over as well? It looks good to me.

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:70]: I think creating a default implementation that blocks is a dangerous short cut. Callers of a method with async in the name that returns a future are not going to expect blocking. Instead I think we should create a new interface called something like AsyncServerSecurityPolicy and support old blocking ServerSecurityPolicies simply by submit()ing them to a ListeningExecutorService (gRPC offload executor?)

**Review by jdcormie — CHANGES_REQUESTED** [review]

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:70]: If for some reason we really need a single ServerSecurityPolicy interface to do everything, perhaps the default checkAuthorizationForServiceAsync() could return null or some sentinel value to indicate this function isn't supported. 

Or perhaps AsyncServerSecurityPolicy extends ServerSecurityPolicy and adds a default implementation of checkAuthorizationForService() that calls get() on the future returned by checkAuthorizationForServiceAsync(). Internally, we could use instanceof to test for AsyncServerSecurityPolicy and prefer that one if available.

Just brainstorming here. Both of these are pretty ugly  ...

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:68]: Supporting `ListenableFuture<Status>` is actually kind of the main goal here, because I want make the framework support a client-provided `SecurityPolicy` that has a slow implementation without blocking the calling thread. An [earlier version](https://github.com/grpc/grpc-java/pull/10596/commits/9a94b59bd4ae1ce55ba38cb9288d267b0ae1cd12) of this PR may provide more context on what I'm aiming for, though @markb74 recommended splitting it to make the review easier.

Also note that this method is package-private. The bulk of the documentation is coming for the public `AsyncSecurityPolicy`. I think there's more value in investing there,  but please let me know if you feel strongly about documenting this one too.

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:70]: See my comment above. There's an `AsyncSecurityPolicy` coming soon ([sneak peak](https://github.com/mateusazis/grpc-java/commit/9a94b59bd4ae1ce55ba38cb9288d267b0ae1cd12)) but it was moved out to make each PR smaller.

Please note that:
- `ServerSecurityPolicy.checkAuthorizationForServiceAsync` currently **does not** block per se; just wraps the original status in a future. So I expect it to be just as performant as the current implementation.
- `BinderTransportSecurity.TransportAuthorizationState.checkAuthorization` **does** block, but, for the time being, should display equivalent performance to the current implementation since async policies are not supported yet (i.e. slow `SecurityPolicy` implementions would block anyway). That method has a TODO which I can't address until I add a few more classes, which I'm holding for now to keep this PR small.

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:68]: I'm not arguing against ListenableFuture. I'm just pointing out that ListenableFutures can complete with any type of exception and the compiler/IDE doesn't help you formalize them with throws declarations or @throws javadoc. As a reader it really isn't clear what it would mean for the returned future to complete with io.grpc.Status.PERMISSION_DENIED.asException() vs completing successfully with Status.PERMISSION_DENIED. 

But thanks for pointing out that this is not actually a public method. Up to you then whether you think formal javadoc is necessary.

File: binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java:136]: Please see go/java-practices/interruption#interruptedexception

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:70]: OK. I get that you are breaking this up into multiple PRs to make each one smaller. And I agree that what you have is no worse performance wise or blocking-wise than what we have today. But I think it's important for each of these small PRs to leave the code in a rational and readable intermediate state. I don't see how that can be done without AsyncSecurityPolicy but I see that's something you've already negotiated with markb74 so maybe I'll just get out of the way.

Comment (markb74) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java]: FWIW, what I didn't like about adding AsyncSecurityPolicy in this PR is that it exposed an API which doesn't do what you'd expect.

This state, while a little ugly, is at least internal only.

In any case, I think we can hold off on merging this until we're ready with the follow up.

File: binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java
Comment (jdcormie) [inline binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java:148]: /** If this is meant to be javadoc?

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (jdcormie) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:71]: Thanks for that addition, although that part is kind of implied by the lack of throws clauses in the declaration.

What I was really asking for is to explain what to expect when it comes to the returned future completing with an exception. How about something like "The returned future is not expected to complete with an Exception -- if it does, that should be interpreted as a failure to make an authorization decision, not the authorization decision itself."

Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:73]: Updated.

File: binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java]: Oops, fixed.

**Review by jdcormie — APPROVED** [review]

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (ejona86) [inline binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:73]: For the operation to be async, should we pass in an executor here? We do that in similar APIs. That would allow this adapter implementation to be async as well.

Not necessary for this PR (since this is all internal API), but something to consider.

File: binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java
Comment (ejona86) [inline binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java:157]: This class is in `internal` package, so it doesn't need any experimental API annotation.

**Review by ejona86 — APPROVED** [review]
Remove the `@ExperimentalApi` and it looks good to go. Nothing's really changed yet.

File: binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java
Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:73]: If we do that, that means the gRPC library is responsible for picking which executor to pass in. And it may not pick the appropriate one without knowing in advance what the implementation intends to do.

What I want here is to give implementers to use the executor of their choice. For instance, Google's `@BlockingExecutor`or `@BackgroundExecutor` depending on the type of blocking call they are issuing. 

From the library's perspective, all that matters is clients trying to be well-behaved and avoiding blocking the gRPC thread, but they are free to choose how to write their async code.

Comment (ejona86) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:73]: gRPC already has an executor for blocking operations:
https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannelBuilder.html#offloadExecutor(java.util.concurrent.Executor)

It is used for DNS and retrieving oauth tokens. That sort of work.

Comment (mateusazis) [reply binder/src/main/java/io/grpc/binder/ServerSecurityPolicy.java:73]: But not all slow calls are the same. go/which-executor, for instance, distinguishes those that scale well (-> `@BlockingExecutor`; eg: network calls) from those that do not (-> `@BackgroundExecutor`; eg: CPU-intensive work, IPC, disk I/O etc.). We don't know which kind will be implemented.

There is a future CL in which I do plan to use an executor internally to handle the resulting `Status` object. For now, however, I think it's cleaner to leave it up to implementers and not suggest any particular executor. They are still welcome to use the same one they passed to the channel builder, but the framework is agnostic to that.

**Pull request merged**
